src/NServiceBus.Persistence.DynamoDB/Serialization/HashSetOfNumberConverter.cs

```diff
@@ -0,0 +1,133 @@
+                                        innerTypeToConvert == typeof(decimal) ||
+                                        innerTypeToConvert == typeof(float);
+
+            return isNumberTypeSupported &&
+                   // This check is fairly expensive so we do it only once we know it is a number type that is supported
+                   typeToConvert.IsAssignableToGenericType(typeof(ISet<>));
+        }
+
+        public override JsonConverter CreateConverter(Type type, JsonSerializerOptions options)
+        {
+            Type valueType = type.GetGenericArguments()[0];
+            var converter = (JsonConverter)Activator.CreateInstance(
+                typeof(SetConverter<,>)
+                    .MakeGenericType(new Type[] { type, valueType }),
+                BindingFlags.Instance | BindingFlags.Public,
+                binder: null,
+                args: null,
+                culture: null)!;
+            return converter;
+
+        }
+        sealed class SetConverter<TSet, TValue> : JsonConverter<TSet>
+            where TSet : ISet<TValue>
+            where TValue : struct
+        {
+            public override TSet? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                if (reader.TokenType != JsonTokenType.StartObject)
+                {
+                    throw new JsonException();
+                }
+
+                reader.Read();
+                if (reader.TokenType != JsonTokenType.PropertyName)
+                {
+                    throw new JsonException();
+                }
+
+                string? propertyName = reader.GetString();
+                if (propertyName != PropertyName)
+                {
+                    throw new JsonException();
+                }
+
+                reader.Read();
+                if (reader.TokenType != JsonTokenType.StartArray)
+                {
+                    throw new JsonException();
+                }
+
+                // Deliberately not passing the options to use the default json serialization behavior
+                var set = JsonSerializer.Deserialize<TSet>(ref reader);
+
+                reader.Read();
+
+                if (reader.TokenType != JsonTokenType.EndObject)
+                {
+                    throw new JsonException();
+                }
+                return set;
+            }
+
+            public override void Write(Utf8JsonWriter writer, TSet value, JsonSerializerOptions options)
+            {
+                writer.WriteStartObject();
+                writer.WritePropertyName(PropertyName);
+                // Deliberately not passing the options to use the default json serialization behavior
+                JsonSerializer.Serialize(writer, value);
+                writer.WriteEndObject();
+            }
+        }
+
+        public static bool TryExtract(JsonProperty property, out List<string?>? numbersAsStrings)
```

Comment:
ok this is a bit of a wild thought and beyond this PR at this point, but what if we wouldn't actually serialize the full collections into json but only a "reference id" instead. When we try to extract the value to create `AttributeValues` from it again, we could try to return the original collection reference on the source object instead, that would save the step to serialize and deserialize collections.

Reply:
Tim and I discussed this and realized that we always need to represent strings as numbers and can't really keep around a reference to the original list. Keeping around an original list would mean we have to have reflection in place and some mechanism to track them. That could easily become slower.